test/nn/test_load_state_dict.py

```diff
@@ -362,7 +365,162 @@ def test_load_state_dict_warn_assign(self):
+                return src
+            else:
+                raise NotImplementedError(f"module_load_to(): unsupported type(dest) {type(dest)}")
+
+        if func is torch.Tensor.module_load_from:
+            return module_load_from(*args, **kwargs)
+        elif func is torch.Tensor.module_load_to:
+            return module_load_to(*args, **kwargs)
+        else:
+            with torch._C.DisableTorchFunctionSubclass():
+                return func(*args, **kwargs)
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+
+        def unwrap(t):
+            return t._data if isinstance(t, MyWrapperLoadTensor) else t
+
+        kwargs = {} if kwargs is None else kwargs
+        out = func(*tree_map(unwrap, args), **tree_map(unwrap, kwargs))
+        if isinstance(out, torch.Tensor):
+            return MyWrapperLoadTensor(out)
+        else:
+            return out
+
+
+class TestLoadStateDictSwap(TestCase):
+    @skipIfCrossRef
+    @skipIfTorchDynamo("Can't swap with dynamo as dynamo installs weakrefs")
+    @wrapSwapTensorsTest(swap=True)
+    def test_swap_subclass(self):
+
+        def _create_model(subclass=None):
+            m = torch.nn.Linear(2, 3, bias=False)
+            m.register_buffer('buf', torch.randn(2, 3))
+            if subclass is not None:
+                m.weight = torch.nn.Parameter(subclass(m.weight))
+                m.buf = subclass(m.buf)
+            return m
+
+        def _test(module_subclass=None, sd_subclass=None):
+            m = _create_model(module_subclass)
+            sd = _create_model(sd_subclass).state_dict()
+            # deepcopy as swap_tensors might destructively modify state dict
+            ref_sd = deepcopy(sd)
+            m.load_state_dict(sd)
+            self.assertEqual(m.weight, ref_sd['weight'])
+            self.assertEqual(m.buf, ref_sd['buf'])
+            self.assertTrue(isinstance(m.weight, torch.nn.Parameter))
+            self.assertTrue(not isinstance(m.buf, torch.nn.Parameter))
+
+            weight_type, buf_type = (torch.nn.Parameter, torch.Tensor)
+            if sd_subclass is not None and module_subclass is not None:
+                # param.module_load_from takes precedence over input.module_load_to
+                weight_type, buf_type = (module_subclass, module_subclass)
+            elif sd_subclass is not None:
+                weight_type, buf_type = (sd_subclass, sd_subclass)
+            elif module_subclass is not None:
+                weight_type, buf_type = (module_subclass, module_subclass)
+            self.assertTrue(type(m.weight) is weight_type)
+            self.assertTrue(type(m.buf) is buf_type)
+
+        subclasses = [None, MyLoadTensor, MyWrapperLoadTensor]
+        for m_s, sd_s in product(subclasses, subclasses):
+            _test(m_s, sd_s)
+
+
+# run tests witsh swap enabled/disabled
+def instantiate_parametrize_with_swap(cls):
```

Comment:
Why do you need this here? Can't it be in the utils or included in the parametrized test generation?

Reply:
Do you mean move this function to a test utils file?

I want to run each test within the parametrized class with two different decorators that contain context managers that set `swap`.

I did not want to include it in the parametrized test generation (`instantiate_parametrized_tests`) as that seemed too "large blast" and I don't think we need to parametrize that many tests with swap True and False

I think using `@parametrize` here would require explicitly using the context manager in every test within the test class which I preferred not to do :P

I agree that my solution here is not clean though, do you know of any utility that would achieve this?

Reply:
Oops ok changed this to a `@swap` `_TestParametrizer` and decorated each function